knowrob_common/prolog/owl.pl

```diff
@@ -709,6 +762,64 @@
 owl_has_direct(S, P, O) :-
 	owl_use_rule(S, P, O).
 
+%% Simplest branch: find an explicitly stored rdf triple (S, P, O)
+owl_has_direct_internal(S, P, O) :-
+	rdf_has(S, P, O).
+
+%% If P is bound to an object property, see if any of its PropertyChain axioms is able to produce explicitly known triples.
+%% ASSUMPTION: no circular PropertyChain axioms (example, P defined as A o B and A defined as P o B)
+owl_has_direct_internal(S, P, O) :-
+        nonvar(P),
+        owl_individual_of(P, owl:'ObjectProperty'),
+        rdf_has(P, owl:propertyChainAxiom, RDFList),
+        rdfs_list_to_prolog_list(RDFList, PropChain),
+% We need this because the PropChain contains a list of either individuals of type owl:'ObjectProperty', 
+% or individuals who are not object properties but have owl:inverseOf some owl:'ObjectProperty',
+% and also we need a simple list of the props appearing in a property chain.
+        kr_get_props(PropChain, Props),
+% Ensure we avoid circular property chains
+        \+ kr_circular_chain(Props, [P]),
+        owl_has_property_chain(S, PropChain, O).
+
+%% No longer needed. We now attempt direct and inverse properties using the owl_has_direct above, and defined a predicate
+%% owl_has_direct_internal that never looks for inverse properties, to avoid infinite backtracking. This fixes the TBD below.
+% owl_has_direct(S, P, O) :-
+%	(   rdf_has(P, owl:inverseOf, P2)
+%	->  true
+%	;   rdf_has(P2, owl:inverseOf, P)
+%	),
+%	rdf_has(O, P2, S).		% TBD: must call owl_has_direct/3
+
+%% Copied above
+% owl_has_direct(S, P, O) :-
+% 	rdfs_individual_of(P, owl:'SymmetricProperty'),
+% 	rdf(O, P, S).
+
+owl_has_direct_internal(S, P, O) :-
+	owl_use_has_value(S, P, O).
+
+owl_has_property_chain(S, PropChain, O) :-
+        nonvar(S),
+        owl_has_property_chain_S2O(S, PropChain, O).
+
+owl_has_property_chain(S, PropChain, O) :-
+        var(S),
+        nonvar(O),
+        reverse(PropChain, PropChainRev),
+        owl_has_property_chain_O2S(O, PropChainRev, S).
+
+owl_has_property_chain_S2O(O, [], O).
+
+owl_has_property_chain_S2O(S, [P|RestChain], O) :-
+        owl_has(S, P, Oi),
+        owl_has_property_chain_S2O(Oi, RestChain, O).
+
+owl_has_property_chain_O2S(S, [], S).
+
+owl_has_property_chain_O2S(O, [P|RestChain], S) :-
+        owl_has(Si, P, O),
```

Comment:
Shouldn't that be `owl_has(O,P,Si)` instead?
Since `P` is inverse relation with `O` as subject.
Then you can remove this clause and just use the above one for both cases.